### PR TITLE
Don't remove a Namespace Index while it still contains AuthorizationPolicy

### PR DESCRIPTION
**Pull request by adleong**
In the inbound policy index, we maintain a policy index per namespace which holds various policy resources for that namespace.  When a per namespace index becomes empty, we remove it.  However, we were not considering authorization policy resources when determining if the index is empty.  This could result in the index being removed even while it contained authorization policy resources, as long as all other resource types did not exist.

This can lead to incorrect inbound policy responses when the per namespace index is recreated, since it will not longer contain the authorization policy.

We update the `is_empty()` function to properly consider authorization policies as well.  We also add some generally useful logging at debug and trace level.

TODO: add test

### CODE REVIEW

File: policy-controller/k8s/index/src/inbound/index.rs [hunk 1]
```diff
@@ -397,6 +397,7 @@ impl kubert::index::IndexNamespacedResource<k8s::Pod> for Index {
             // watches will complete.  No other parts of the index need to be
             // updated.
             if ns.get_mut().pods.by_name.remove(&name).is_some() && ns.get().is_empty() {
+                tracing::debug!(namespace = ns.key(), "removing empty namespace index");
```
Comment (olix0r) [inline policy-controller/k8s/index/src/inbound/index.rs:400]:
```suggestion
                tracing::debug!(namespace = ns.key(), "Removing empty namespace index");
```

File: policy-controller/k8s/index/src/inbound/index.rs [hunk 2]
```diff
@@ -622,6 +623,7 @@ impl kubert::index::IndexNamespacedResource<k8s::policy::AuthorizationPolicy> fo
 
     fn delete(&mut self, ns: String, ap: String) {
         let _span = info_span!("delete", %ns, %ap).entered();
+        tracing::trace!(%ap, "delete");
```
Comment (olix0r) [inline policy-controller/k8s/index/src/inbound/index.rs:626]:
```suggestion
        tracing::trace!(name = %ap, "Delete");
```

File: policy-controller/k8s/index/src/inbound/index.rs [hunk 3]
```diff
@@ -634,6 +636,7 @@ impl kubert::index::IndexNamespacedResource<k8s::policy::AuthorizationPolicy> fo
     ) {
         let _span = info_span!("reset");
 
+        tracing::trace!(?deleted, ?policies, "reset");
```
Comment (olix0r) [inline policy-controller/k8s/index/src/inbound/index.rs:639]:
```suggestion
        tracing::trace!(?deleted, ?policies, "Reset");
```

File: policy-controller/k8s/index/src/inbound/index.rs [hunk 4]
```diff
@@ -902,6 +905,7 @@ impl NamespaceIndex {
         if let Entry::Occupied(mut ns) = self.by_ns.entry(namespace) {
             if f(ns.get_mut()) {
                 if ns.get().is_empty() {
+                    tracing::debug!(namespace = ns.key(), "removing empty namespace index");
```
Comment (olix0r) [inline policy-controller/k8s/index/src/inbound/index.rs:908]:
```suggestion
                    tracing::debug!(namespace = ns.key(), "Removing empty namespace index");
```

**Review by olix0r — COMMENTED** [review]
some logging consistency nits...

How hard is it to write a test for this?

**Review by olix0r — APPROVED** [review]

**Review by mateiidavid — APPROVED** [review]
🚢

**Pull request merged**